File: python/Solution_Sword_04.py
# !/usr/bin/python
# coding: utf8
# Time: 2020/9/24 上午9:51
# Author: Liam
# E-mail: luyu.real@qq.com
# Software: PyCharm
class Solution:
    def findNumberIn2DArray(self, matrix, target):
        if matrix is None or len(matrix) == 0:
            return False
        m, n = len(matrix), len(matrix[0])
        row, col = 0, n - 1
        while row < m and col >= 0:
            if target > matrix[row][col]:
                row += 1
            elif target < matrix[row][col]:
                col -= 1
            else:
                return True
        return False


if __name__ == '__main__':
    print(Solution().findNumberIn2DArray([
        [1, 4, 7, 11, 15],
        [2, 5, 8, 12, 19],
        [3, 6, 9, 16, 22],
        [10, 13, 14, 17, 24],
        [18, 21, 23, 26, 30]
    ], 20))

File: java/src/Solution_Sword_04.java
public class Solution_Sword_04 {
    public boolean findNumberIn2DArray(int[][] matrix, int target) {
        if (matrix == null || matrix.length == 0) {
            return false;
        }
        int m = matrix.length;
        int n = matrix[0].length;
        int row = 0, col = n-1;
        while (row < m && col >= 0) {
            if (target > matrix[row][col]) {
                ++row;
            } else if (target < matrix[row][col]) {
                --col;
            } else {
                return true;
            }
        }

        return false;
    }
}
